tools/coq_makefile.ml

```diff
@@ -151,6 +151,51 @@ let generate_makefile oc conf_file local_file local_late_file dep_file args proj
     ] in
   output_string oc s
 
+let generate_meta_file p =
+  try
+    match p.meta_file with
+    | Absent -> p
+    | Generate proj ->
+        let cmname = List.map (fun { thing } -> thing) (p.mllib_files @ p.mlpack_files) in
+        let dir, cmname =
+          match cmname with
+          | [] -> Printf.eprintf "In order to generate a META file one needs an .mlpack or .mllib file"; exit 1
+          | [x] -> Filename.dirname x, Filename.(basename @@ chop_extension x)
+          | _ -> Printf.eprintf "Automatic META generation only works for one .mlpack or .mllib file, since you have more you need to write the META file by hand"; exit 1 in
+        let f = "META." ^ proj in
+        let oc = open_out f in
+        let meta = {|
+package "plugin" (
+  directory = "@@DIR@@"
+  requires = "coq-core.plugins.ltac"
+  archive(byte) = "@@CMNAME@@.cma"
+  archive(native) = "@@CMNAME@@.cmxa"
+  plugin(byte) = "@@CMNAME@@.cma"
+  plugin(native) = "@@CMNAME@@.cmxs"
+)
+directory = "."
+|}
+        in
+        let meta = Str.global_replace (Str.regexp_string "@@DIR@@") dir meta in
+        let meta = Str.global_replace (Str.regexp_string "@@CMNAME@@") cmname meta in
```

Comment:
might be nicer using printf:
~~~ocaml
let meta : _ format = {|
package "plugin" (
  directory = "%s"
  requires = "coq-core.plugins.ltac"
  archive(byte) = "%s.cma"
  archive(native) = "%s.cmxa"
  plugin(byte) = "%s.cma"
  plugin(native) = "%s.cmxs"
)
directory = "."
|}
in
let meta : string = Printf.sprintf meta dir cmname cmname cmname cmname in
~~~

Reply:
thanks, I guess I tried but forgot the type annotation and boom...